File: src/main/java/com/almightyalpaca/jetbrains/plugins/discord/listeners/DocumentListener.java
/*
 * Copyright 2017-2018 Aljoscha Grebe
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.almightyalpaca.jetbrains.plugins.discord.listeners;

import com.almightyalpaca.jetbrains.plugins.discord.components.ProjectComponent;
import com.almightyalpaca.jetbrains.plugins.discord.debug.Logger;
import com.almightyalpaca.jetbrains.plugins.discord.debug.LoggerFactory;
import com.intellij.openapi.editor.Document;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.editor.EditorFactory;
import com.intellij.openapi.editor.event.DocumentEvent;
import com.intellij.openapi.fileEditor.FileDocumentManager;
import com.intellij.openapi.project.Project;
import org.jetbrains.annotations.NotNull;

@SuppressWarnings("Duplicates")
public class DocumentListener implements com.intellij.openapi.editor.event.DocumentListener
{
    @NotNull
    private static final Logger LOG = LoggerFactory.getLogger(DocumentListener.class);

    @Override
    public void beforeDocumentChange(DocumentEvent event) {}

    @Override
    public void documentChanged(DocumentEvent event)
    {
        LOG.trace("DocumentListener#documentChanged({})", event.getDocument());

        Document document = event.getDocument();
        Editor[] editors = EditorFactory.getInstance().getEditors(document);
        FileDocumentManager documentManager = FileDocumentManager.getInstance();

        for (Editor editor : editors)
        {
            Project project = editor.getProject();
            ProjectComponent component = ProjectComponent.getInstance(project);

            if (component != null)
                component.updateTimeAccessed(documentManager.getFile(document));
        }
    }
}

File: plugin/src/main/kotlin/com/almightyalpaca/jetbrains/plugins/discord/plugin/listeners/DocumentListener.kt
/*
 * Copyright 2017-2019 Aljoscha Grebe
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.almightyalpaca.jetbrains.plugins.discord.plugin.listeners

import com.almightyalpaca.jetbrains.plugins.discord.plugin.components.ApplicationComponent
import com.almightyalpaca.jetbrains.plugins.discord.plugin.logging.Logging
import com.intellij.openapi.editor.EditorFactory
import com.intellij.openapi.editor.event.DocumentEvent
import com.intellij.openapi.fileEditor.FileDocumentManager
import com.intellij.openapi.project.ProjectManagerListener
import java.time.OffsetDateTime

class DocumentListener : com.intellij.openapi.editor.event.DocumentListener, ProjectManagerListener {

    override fun beforeDocumentChange(event: DocumentEvent) {}

    override fun documentChanged(event: DocumentEvent) {
        log { "DocumentListener#documentChanged($event)" }

        val document = event.document
        val editors = EditorFactory.getInstance().getEditors(document)
        val file = FileDocumentManager.getInstance().getFile(document)

        ApplicationComponent.instance.app {
            for (editor in editors) {
                update(editor.project) {
                    update(file) {
                        accessedAt = OffsetDateTime.now()
                    }
                }
            }
        }
    }

    companion object : Logging()
}
